Source/WebCore/dom/InternalObserverFirst.cpp

```diff
@@ -0,0 +1,111 @@
+*    documentation and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+* THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+* PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+* BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+* THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "config.h"
+#include "InternalObserverFirst.h"
+
+#include "AbortController.h"
+#include "AbortSignal.h"
+#include "Exception.h"
+#include "ExceptionCode.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverFirst final : public InternalObserver {
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverFirst(context, signal, WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->resolve<IDLAny>(value);
+        (Ref { m_signal })->signalAbort(JSC::jsUndefined());
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        (Ref { m_promise })->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor&) const final
+    {
+    }
+
+    InternalObserverFirst(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_signal(signal)
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    Ref<AbortSignal> m_signal;
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Ref<Observable> observable, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    Ref signal = AbortSignal::create(&context);
+
+    Vector<Ref<AbortSignal>> dependentSignals = { signal };
```

Comment:
`WTFMove(signal)` or do away with the local variable entirely and do it on one line

Reply:
Ill make it so, are you saying:

```
Vector<Ref<AbortSignal>> dependentSignals = { AbortSignal::create(&context) };
``

Reply:
Actually a good catch here mate. Really we want to pass this `signal` down into `InternalObserverFirst::create`. BUT! When I do that, [this promise hangs](https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js#L66), and times out. 

Through a lot of debugging, our `dependentSignal->addAlgorithm` never fires. I did verify that the [`m_dependentSignals`](https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/Source/WebCore/dom/AbortSignal.cpp#L122)'s `computedSize()` does increment for this test specifically. But then right before [this line](https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/Source/WebCore/dom/AbortSignal.cpp#L137), that count is zero.

I suspect its lifetimes? Do I need to WTFMove something?

Note; the current HEAD diff here has that `Ref signal = AbortSignal::create(&context);` removed, as per your suggestion to inline it. But think that is wrong [as per the spec, point 3](https://wicg.github.io/observable/#dom-observable-first). 

Though the tests pass, I think are by chance — but does not make logical sense imo.

Reply:
I read Darin's suggestion as pick one of these two ways to populate the Vector:

```C++
Ref signal = AbortSignal::create(&context);
Vector<Ref<AbortSignal>> dependentSignals = { WTFMove(signal) };
```

or 

```C++
Vector<Ref<AbortSignal>> dependentSignals = { AbortSignal::create(&context) };
```

To be honest I don't know enough about this API to know why your current branch has a bug here but you did change the composition of `dependentSignals` by initializing it as an empty array now.

Reply:
Yes @rreno understood me correctly. I wasn't offering deeper insight about what is correct or not with the code.